marshaller.go

```diff
@@ -0,0 +1,64 @@
+package commando
+
+import (
+	"encoding/csv"
+	"reflect"
+)
+
+// Marshaller is a CSV to struct marshaller.
+type Marshaller struct {
+	config *validConfig
+	writer *csv.Writer
+}
+
+// NewMarshaller is a convenience function which allocates and
+// returns a new Marshaller.
+func NewMarshaller(holder interface{}, writer *csv.Writer) (*Marshaller, error) {
+	return (&Config{Holder: holder}).NewMarshaller(writer)
+}
+
+// NewMarshaller creates a marshaller from a csv.Writer.
+func (c *Config) NewMarshaller(writer *csv.Writer) (*Marshaller, error) {
+	vc, err := c.validate(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &Marshaller{
+		writer: writer,
+		config: vc,
+	}
+
+	if err := m.writeHeaders(); err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (m *Marshaller) writeHeaders() error {
+	return m.writer.Write(m.config.structInfo.headers())
+}
+
+func (m *Marshaller) Write(record interface{}) error {
+
+	inValue, inType := getConcreteReflectValueAndType(record) // Get the concrete type
+	if err := ensureStructOrPtr(inType); err != nil {
+		return err
+	}
+	inInnerWasPointer := inType.Kind() == reflect.Ptr
```

Comment:
the comparison is probably trivial, so maybe just store the result here for comparisons later. there are other ways Elem() can explode IIRC

Reply:
what I mean is, you later pass this value to a function as reduced to bool, so it can be processed. what I'm suggesting is that there may be other reasons for this that may not be as easy to express this way, so you may just want to store the result of `Kind()` and process that in your function instead.

Reply:
and tbh, I think Kind is just a struct deref and likely all this work is done at TypeOf/ValueOf time; makes me wonder if the call overhead is really worth this optimization.